File: scripts/android/lib/patch_gradle_files.py
#!/usr/bin/env python3
"""Utilities to normalize generated Gradle build files for Android tests."""
from __future__ import annotations

import argparse
import pathlib
import re
from typing import Tuple

REPOSITORIES_BLOCK = """\
repositories {
    google()
    mavenCentral()
}
"""


def _ensure_repositories(content: str) -> Tuple[str, bool]:
    """Ensure a repositories block exists with google() and mavenCentral()."""
    pattern = re.compile(r"(?ms)^\s*repositories\s*\{.*?\}")
    match = pattern.search(content)
    block_added = False

    if not match:
        # Append a canonical repositories block to the end of the file.
        if not content.endswith("\n"):
            content += "\n"
        content += REPOSITORIES_BLOCK
        return content, True

    block = match.group(0)
    if "google()" not in block or "mavenCentral()" not in block:
        lines = block.splitlines()
        header = lines[0]
        body = [line for line in lines[1:-1] if line.strip()]
        if "google()" not in block:
            body.append("    google()")
        if "mavenCentral()" not in block:
            body.append("    mavenCentral()")
        new_block = "\n".join([header, *sorted(set(body)), lines[-1]])
        content = content[: match.start()] + new_block + content[match.end() :]
        block_added = True
    return content, block_added


def _ensure_android_sdk(content: str, compile_sdk: int, target_sdk: int) -> Tuple[str, bool]:
    changed = False

    def insert_or_replace(pattern: str, repl: str, search_scope: str) -> Tuple[str, bool]:
        nonlocal content
        match = re.search(pattern, content, re.MULTILINE)
        if match:
            new_content = re.sub(pattern, repl, content, count=1, flags=re.MULTILINE)
            if new_content != content:
                content = new_content
                return content, True
            return content, False
        anchor = re.search(search_scope, content, re.MULTILINE)
        if not anchor:
            return content, False
        start = anchor.end()
        content = content[:start] + f"\n{repl}" + content[start:]
        return content, True

    # Ensure android block exists
    if re.search(r"(?m)^\s*android\s*\{", content) is None:
        if not content.endswith("\n"):
            content += "\n"
        content += (
            "\nandroid {\n"
            f"    compileSdkVersion {compile_sdk}\n"
            "    defaultConfig {\n"
            f"        targetSdkVersion {target_sdk}\n"
            "    }\n}"
        )
        return content, True

    new_content, changed_once = insert_or_replace(
        pattern=rf"(?m)^\s*compileSdkVersion\s+\d+",
        repl=f"    compileSdkVersion {compile_sdk}",
        search_scope=r"(?m)^\s*android\s*\{",
    )
    changed = changed or changed_once

    default_config = re.search(r"(?ms)^\s*defaultConfig\s*\{.*?^\s*\}", content)
    if default_config:
        block = default_config.group(0)
        if re.search(r"(?m)^\s*targetSdkVersion\s+\d+", block):
            updated = re.sub(
                r"(?m)^\s*targetSdkVersion\s+\d+",
                f"        targetSdkVersion {target_sdk}",
                block,
            )
        else:
            updated = re.sub(r"{", "{\n        targetSdkVersion %d" % target_sdk, block, count=1)
        if updated != block:
            content = content[: default_config.start()] + updated + content[default_config.end() :]
            changed = True
    else:
        content, changed_once = insert_or_replace(
            pattern=r"(?ms)(^\s*android\s*\{)",
            repl="    defaultConfig {\n        targetSdkVersion %d\n    }" % target_sdk,
            search_scope=r"(?m)^\s*android\s*\{",
        )
        changed = changed or changed_once

    return content, changed


def _ensure_instrumentation_runner(content: str) -> Tuple[str, bool]:
    runner = "androidx.test.runner.AndroidJUnitRunner"
    if runner in content:
        return content, False
    changed = False
    pattern = re.compile(r"(?m)^\s*testInstrumentationRunner\s*\".*?\"\s*$")
    if pattern.search(content):
        new_content = pattern.sub(f"        testInstrumentationRunner \"{runner}\"", content)
        return new_content, True
    default_config = re.search(r"(?ms)^\s*defaultConfig\s*\{", content)
    if default_config:
        insert_point = default_config.end()
        content = (
            content[:insert_point]
            + f"\n        testInstrumentationRunner \"{runner}\""
            + content[insert_point:]
        )
        changed = True
    else:
        android_block = re.search(r"(?ms)^\s*android\s*\{", content)
        if android_block:
            insert_point = android_block.end()
            snippet = (
                "\n    defaultConfig {\n"
                f"        testInstrumentationRunner \"{runner}\"\n"
                "    }"
            )
            content = content[:insert_point] + snippet + content[insert_point:]
            changed = True
    return content, changed


def _remove_legacy_use_library(content: str) -> Tuple[str, bool]:
    new_content, count = re.subn(
        r"(?m)^\s*useLibrary\s+'android\.test\.(?:base|mock|runner)'\s*$",
        "",
        content,
    )
    return new_content, bool(count)


def _ensure_test_dependencies(content: str) -> Tuple[str, bool]:
    module_view = re.sub(r"(?ms)^\s*(buildscript|pluginManagement)\s*\{.*?^\s*\}", "", content)
    uses_modern = re.search(
        r"(?m)^\s*(implementation|api|testImplementation|androidTestImplementation)\b",
        module_view,
    )
    configuration = "androidTestImplementation" if uses_modern else "androidTestCompile"
    dependencies = [
        "androidx.test.ext:junit:1.1.5",
        "androidx.test:runner:1.5.2",
        "androidx.test:core:1.5.0",
        "androidx.test.services:storage:1.4.2",
    ]
    missing = [dep for dep in dependencies if dep not in module_view]
    if not missing:
        return content, False
    block = "\n\ndependencies {\n" + "".join(
        f"    {configuration} \"{dep}\"\n" for dep in missing
    ) + "}\n"
    if not content.endswith("\n"):
        content += "\n"
    return content + block, True


def patch_app_build_gradle(path: pathlib.Path, compile_sdk: int, target_sdk: int) -> bool:
    text = path.read_text(encoding="utf-8")
    changed = False

    for transform in (
        lambda c: _ensure_android_sdk(c, compile_sdk, target_sdk),
        _ensure_instrumentation_runner,
        _remove_legacy_use_library,
        _ensure_test_dependencies,
    ):
        text, modified = transform(text)
        changed = changed or modified

    if changed:
        path.write_text(text if text.endswith("\n") else text + "\n", encoding="utf-8")
    return changed


def patch_root_build_gradle(path: pathlib.Path) -> bool:
    text = path.read_text(encoding="utf-8")
    text, changed = _ensure_repositories(text)
    if changed:
        path.write_text(text if text.endswith("\n") else text + "\n", encoding="utf-8")
    return changed


def main() -> int:
    parser = argparse.ArgumentParser(description="Normalize Gradle build files")
    parser.add_argument("--root", required=True, type=pathlib.Path, help="Path to root build.gradle")
    parser.add_argument("--app", required=True, type=pathlib.Path, help="Path to app/build.gradle")
    parser.add_argument("--compile-sdk", type=int, default=33)
    parser.add_argument("--target-sdk", type=int, default=33)
    args = parser.parse_args()

    modified_root = patch_root_build_gradle(args.root)
    modified_app = patch_app_build_gradle(args.app, args.compile_sdk, args.target_sdk)

    if modified_root:
        print(f"Patched {args.root}")
    if modified_app:
        print(f"Patched {args.app}")
    if not (modified_root or modified_app):
        print("Gradle files already normalized")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/android/lib/PatchGradleFiles.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PatchGradleFiles {
    private static final String REPOSITORIES_BLOCK = """
            repositories {
                google()
                mavenCentral()
            }
            """.stripTrailing();

    private static final Pattern REPOSITORIES_PATTERN = Pattern.compile("(?ms)^\\s*repositories\\s*\\{.*?\\}");

    private static final Pattern ANDROID_BLOCK_PATTERN = Pattern.compile("(?m)^\\s*android\\s*\\{");
    private static final Pattern DEFAULT_CONFIG_PATTERN = Pattern.compile("(?ms)^\\s*defaultConfig\\s*\\{.*?^\\s*\\}");
    private static final Pattern DEFAULT_CONFIG_HEADER_PATTERN = Pattern.compile("(?ms)^\\s*defaultConfig\\s*\\{");
    private static final Pattern COMPILE_SDK_PATTERN = Pattern.compile("(?m)^\\s*compileSdkVersion\\s+\\d+");
    private static final Pattern TARGET_SDK_PATTERN = Pattern.compile("(?m)^\\s*targetSdkVersion\\s+\\d+");
    private static final Pattern TEST_INSTRUMENTATION_PATTERN = Pattern.compile("(?m)^\\s*testInstrumentationRunner\\s*\".*?\"\\s*$");
    private static final Pattern USE_LIBRARY_PATTERN = Pattern.compile("(?m)^\\s*useLibrary\\s+'android\\.test\\.(?:base|mock|runner)'\\s*$");
    private static final Pattern DEPENDENCY_PATTERN = Pattern.compile("(?m)^\\s*(implementation|api|testImplementation|androidTestImplementation)\\b");

    public static void main(String[] args) throws Exception {
        Arguments arguments = Arguments.parse(args);
        if (arguments == null) {
            System.exit(2);
            return;
        }

        boolean modifiedRoot = patchRootBuildGradle(arguments.root);
        boolean modifiedApp = patchAppBuildGradle(arguments.app, arguments.compileSdk, arguments.targetSdk);

        if (modifiedRoot) {
            System.out.println("Patched " + arguments.root);
        }
        if (modifiedApp) {
            System.out.println("Patched " + arguments.app);
        }
        if (!modifiedRoot && !modifiedApp) {
            System.out.println("Gradle files already normalized");
        }
    }

    private static boolean patchRootBuildGradle(Path path) throws IOException {
        String content = Files.readString(path, StandardCharsets.UTF_8);
        Matcher matcher = REPOSITORIES_PATTERN.matcher(content);
        if (!matcher.find()) {
            if (!content.endsWith("\n")) {
                content += "\n";
            }
            content += REPOSITORIES_BLOCK;
            Files.writeString(path, ensureTrailingNewline(content), StandardCharsets.UTF_8);
            return true;
        }

        String block = matcher.group();
        boolean changed = false;
        if (!block.contains("google()") || !block.contains("mavenCentral()")) {
            String[] lines = block.split("\n");
            java.util.LinkedHashSet<String> body = new java.util.LinkedHashSet<>();
            for (int i = 1; i < lines.length - 1; i++) {
                String line = lines[i].trim();
                if (!line.isEmpty()) {
                    body.add("    " + line.trim());
                }
            }
            body.add("    google()");
            body.add("    mavenCentral()");
            StringBuilder newBlock = new StringBuilder();
            newBlock.append(lines[0]).append('\n');
            for (String line : body) {
                newBlock.append(line).append('\n');
            }
            newBlock.append(lines[lines.length - 1]);
            content = content.substring(0, matcher.start()) + newBlock + content.substring(matcher.end());
            changed = true;
        }

        if (changed) {
            Files.writeString(path, ensureTrailingNewline(content), StandardCharsets.UTF_8);
        }
        return changed;
    }

    private static boolean patchAppBuildGradle(Path path, int compileSdk, int targetSdk) throws IOException {
        String content = Files.readString(path, StandardCharsets.UTF_8);
        boolean changed = false;

        Result r = ensureAndroidBlock(content, compileSdk, targetSdk);
        content = r.content();
        changed |= r.changed();

        r = ensureInstrumentationRunner(content);
        content = r.content();
        changed |= r.changed();

        r = removeLegacyUseLibrary(content);
        content = r.content();
        changed |= r.changed();

        r = ensureTestDependencies(content);
        content = r.content();
        changed |= r.changed();

        r = ensureJacocoConfiguration(content);
        content = r.content();
        changed |= r.changed();

        if (changed) {
            Files.writeString(path, ensureTrailingNewline(content), StandardCharsets.UTF_8);
        }
        return changed;
    }

    private static Result ensureAndroidBlock(String content, int compileSdk, int targetSdk) {
        Matcher androidBlockMatcher = ANDROID_BLOCK_PATTERN.matcher(content);
        if (!androidBlockMatcher.find()) {
            if (!content.endsWith("\n")) {
                content += "\n";
            }
            String block = "\nandroid {\n" +
                    "    compileSdkVersion " + compileSdk + "\n" +
                    "    defaultConfig {\n" +
                    "        targetSdkVersion " + targetSdk + "\n" +
                    "    }\n}";
            return new Result(content + block, true);
        }

        boolean changed = false;
        Matcher compileMatcher = COMPILE_SDK_PATTERN.matcher(content);
        if (compileMatcher.find()) {
            String replacement = "    compileSdkVersion " + compileSdk;
            String newContent = compileMatcher.replaceFirst(replacement);
            if (!newContent.equals(content)) {
                content = newContent;
                changed = true;
            }
        } else {
            Matcher insertMatcher = ANDROID_BLOCK_PATTERN.matcher(content);
            if (insertMatcher.find()) {
                int pos = insertMatcher.end();
                content = content.substring(0, pos) + "\n    compileSdkVersion " + compileSdk + content.substring(pos);
                changed = true;
            }
        }

        Matcher defaultConfigMatcher = DEFAULT_CONFIG_PATTERN.matcher(content);
        if (defaultConfigMatcher.find()) {
            String block = defaultConfigMatcher.group();
            Matcher targetMatcher = TARGET_SDK_PATTERN.matcher(block);
            String replacement = "        targetSdkVersion " + targetSdk;
            String updated;
            if (targetMatcher.find()) {
                updated = targetMatcher.replaceFirst(replacement);
            } else {
                int brace = block.indexOf('{');
                if (brace >= 0) {
                    updated = block.substring(0, brace + 1) + "\n" + replacement + block.substring(brace + 1);
                } else {
                    updated = block;
                }
            }
            if (!updated.equals(block)) {
                content = content.substring(0, defaultConfigMatcher.start()) + updated + content.substring(defaultConfigMatcher.end());
                changed = true;
            }
        } else {
            Matcher insertMatcher = ANDROID_BLOCK_PATTERN.matcher(content);
            if (insertMatcher.find()) {
                int pos = insertMatcher.end();
                String snippet = "\n    defaultConfig {\n        targetSdkVersion " + targetSdk + "\n    }";
                content = content.substring(0, pos) + snippet + content.substring(pos);
                changed = true;
            }
        }

        return new Result(content, changed);
    }

    private static Result ensureInstrumentationRunner(String content) {
        String runner = "androidx.test.runner.AndroidJUnitRunner";
        if (content.contains(runner)) {
            return new Result(content, false);
        }
        Matcher matcher = TEST_INSTRUMENTATION_PATTERN.matcher(content);
        if (matcher.find()) {
            String replacement = "        testInstrumentationRunner \"" + runner + "\"";
            String newContent = matcher.replaceAll(replacement);
            return new Result(newContent, !newContent.equals(content));
        }

        Matcher defaultConfigHeaderMatcher = DEFAULT_CONFIG_HEADER_PATTERN.matcher(content);
        if (defaultConfigHeaderMatcher.find()) {
            int pos = defaultConfigHeaderMatcher.end();
            String snippet = "\n        testInstrumentationRunner \"" + runner + "\"";
            content = content.substring(0, pos) + snippet + content.substring(pos);
            return new Result(content, true);
        }

        Matcher androidMatcher = ANDROID_BLOCK_PATTERN.matcher(content);
        if (androidMatcher.find()) {
            int pos = androidMatcher.end();
            String snippet = "\n    defaultConfig {\n        testInstrumentationRunner \"" + runner + "\"\n    }";
            content = content.substring(0, pos) + snippet + content.substring(pos);
            return new Result(content, true);
        }
        return new Result(content, false);
    }

    private static Result removeLegacyUseLibrary(String content) {
        Matcher matcher = USE_LIBRARY_PATTERN.matcher(content);
        String newContent = matcher.replaceAll("");
        return new Result(newContent, !newContent.equals(content));
    }

    private static Result ensureTestDependencies(String content) {
        String moduleView = content.replaceAll("(?ms)^\\s*(buildscript|pluginManagement)\\s*\\{.*?^\\s*\\}", "");
        boolean usesModern = DEPENDENCY_PATTERN.matcher(moduleView).find();
        String configuration = usesModern ? "androidTestImplementation" : "androidTestCompile";
        String[] dependencies = {
                "androidx.test.ext:junit:1.1.5",
                "androidx.test:runner:1.5.2",
                "androidx.test:core:1.5.0",
                "androidx.test.services:storage:1.4.2"
        };
        boolean missing = false;
        for (String dep : dependencies) {
            if (!moduleView.contains(dep)) {
                missing = true;
                break;
            }
        }
        if (!missing) {
            return new Result(content, false);
        }
        StringBuilder block = new StringBuilder();
        block.append("\n\ndependencies {\n");
        for (String dep : dependencies) {
            if (!moduleView.contains(dep)) {
                block.append("    ").append(configuration).append(" \"").append(dep).append("\"\n");
            }
        }
        block.append("}\n");
        if (!content.endsWith("\n")) {
            content += "\n";
        }
        return new Result(content + block, true);
    }

    private static Result ensureJacocoConfiguration(String content) {
        if (content.contains("jacocoAndroidReport")) {
            return new Result(content, false);
        }

        String jacocoBlock = """
apply plugin: 'jacoco'

android {
    buildTypes {
        debug {
            testCoverageEnabled true
        }
    }
}

jacoco {
    toolVersion = "0.8.11"
}

tasks.register("jacocoAndroidReport", JacocoReport) {
    group = "verification"
    description = "Generates Jacoco coverage report for the debug variant"
    dependsOn("connectedDebugAndroidTest")

    reports {
        xml.required = true
        html.required = true
        html.outputLocation = layout.buildDirectory.dir("reports/jacoco/jacocoAndroidReport/html")
    }

    def excludes = [
            '**/R.class',
            '**/R$*.class',
            '**/BuildConfig.*',
            '**/Manifest*.*',
            '**/*Test*.*',
            '**/androidx/**/*',
            '**/com/google/**/*'
    ]

    def javaClasses = fileTree(dir: "$buildDir/intermediates/javac/debug/classes", exclude: excludes)
    def kotlinClasses = fileTree(dir: "$buildDir/tmp/kotlin-classes/debug", exclude: excludes)

    classDirectories.setFrom(files(javaClasses, kotlinClasses).asFileTree.matching {
        include 'com/codename1/impl/android/**'
    })

    sourceDirectories.setFrom(files("src/main/java"))

    executionData.setFrom(fileTree(dir: "$buildDir", includes: [
            "outputs/code_coverage/debugAndroidTest/connected/*coverage.ec",
            "outputs/code_coverage/connected/*coverage.ec",
            "jacoco/testDebugUnitTest.exec",
            "outputs/unit_test_code_coverage/debugUnitTest/testDebugUnitTest.exec"
    ]))
}
""".stripTrailing();

        return new Result(ensureTrailingNewline(content) + "\n" + jacocoBlock + "\n", true);
    }

    private static String ensureTrailingNewline(String content) {
        return content.endsWith("\n") ? content : content + "\n";
    }

    private record Result(String content, boolean changed) {
    }

    private static class Arguments {
        final Path root;
        final Path app;
        final int compileSdk;
        final int targetSdk;

        Arguments(Path root, Path app, int compileSdk, int targetSdk) {
            this.root = root;
            this.app = app;
            this.compileSdk = compileSdk;
            this.targetSdk = targetSdk;
        }

        static Arguments parse(String[] args) {
            Path root = null;
            Path app = null;
            int compileSdk = 33;
            int targetSdk = 33;
            for (int i = 0; i < args.length; i++) {
                String arg = args[i];
                switch (arg) {
                    case "--root" -> {
                        if (i + 1 >= args.length) {
                            System.err.println("Missing value for --root");
                            return null;
                        }
                        root = Path.of(args[++i]);
                    }
                    case "--app" -> {
                        if (i + 1 >= args.length) {
                            System.err.println("Missing value for --app");
                            return null;
                        }
                        app = Path.of(args[++i]);
                    }
                    case "--compile-sdk" -> {
                        if (i + 1 >= args.length) {
                            System.err.println("Missing value for --compile-sdk");
                            return null;
                        }
                        compileSdk = Integer.parseInt(args[++i]);
                    }
                    case "--target-sdk" -> {
                        if (i + 1 >= args.length) {
                            System.err.println("Missing value for --target-sdk");
                            return null;
                        }
                        targetSdk = Integer.parseInt(args[++i]);
                    }
                    default -> {
                        System.err.println("Unknown argument: " + arg);
                        return null;
                    }
                }
            }
            if (root == null || app == null) {
                System.err.println("--root and --app are required");
                return null;
            }
            return new Arguments(root, app, compileSdk, targetSdk);
        }
    }
}
